test/e2e/cluster/cluster.go

```diff
@@ -0,0 +1,323 @@
+
+		// stop slave
+		f.ExecSQLOnNode(cluster, 1, "root", pw, "STOP SLAVE;")
+
+		// expect node to be removed from service and status to be updated
+		f.NodeEventuallyCondition(cluster, cluster.GetPodHostname(1), api.NodeConditionReplicating, core.ConditionFalse, 30*time.Second)
+		// node 1 should not be in healty service
+		testClusterEndpoints(f, cluster, []int{0}, []int{0}, "after stop slave")
+	})
+
+	It("slave latency", func() {
+		// create a cluster
+		pw := "rootPassword6"
+		secret := tutil.NewClusterSecret("latency", f.Namespace.Name, pw)
+		_, err := f.ClientSet.CoreV1().Secrets(f.Namespace.Name).Create(secret)
+		Expect(err).NotTo(HaveOccurred(), "Failed to create secret '%s'", secret.Name)
+
+		cluster := tutil.NewCluster("latency", f.Namespace.Name)
+		cluster.Spec.Replicas = 2
+		one := int64(1)
+		cluster.Spec.MaxSlaveLatency = &one
+
+		cluster, err = f.MyClientSet.MysqlV1alpha1().MysqlClusters(f.Namespace.Name).Create(cluster)
+		Expect(err).NotTo(HaveOccurred(), "Failed to create cluster: '%s'", cluster.Name)
+
+		// test cluster to be ready
+		testCreateACluster(f, cluster, "after cluster creation")
+		testClusterIsInOrchestartor(f, cluster, "after cluster creation")
+
+		cluster, err = f.MyClientSet.MysqlV1alpha1().MysqlClusters(f.Namespace.Name).Get(cluster.Name, meta.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "Failed to get cluster %s", cluster.Name)
+
+		// set delayed replication
+		f.ExecSQLOnNode(cluster, 1, "root", pw, "STOP SLAVE; CHANGE MASTER TO MASTER_DELAY = 100; START SLAVE;")
+
+		// expect node to be marked as lagged and removed from service
+		f.NodeEventuallyCondition(cluster, cluster.GetPodHostname(1), api.NodeConditionLagged, core.ConditionTrue, 20*time.Second)
+		// node 1 should not be in healty service
+		testClusterEndpoints(f, cluster, []int{0}, []int{0}, "after delayed slave")
+	})
+
+})
+
+func testCreateACluster(f *framework.Framework, cluster *api.MysqlCluster, where string) {
+	timeout := time.Duration(cluster.Spec.Replicas) * f.Timeout
+	By(fmt.Sprintf("Test cluster is ready: %s; timeout=%v", where, timeout))
+
+	// wait for pods to be ready
+	Eventually(func() int {
+		cluster, _ := f.MyClientSet.MysqlV1alpha1().MysqlClusters(f.Namespace.Name).Get(cluster.Name, meta.GetOptions{})
+		return cluster.Status.ReadyNodes
+	}, timeout, POLLING).Should(Equal(int(cluster.Spec.Replicas)), "Not ready replicas of cluster '%s'", cluster.Name)
+
+	f.ClusterEventuallyCondition(cluster, api.ClusterConditionReady, core.ConditionTrue, f.Timeout)
+	// TODO: investigate way sometime exists failover ACK even to a newly created cluster.
+	// f.ClusterEventuallyCondition(cluster, api.ClusterConditionFailoverAck, core.ConditionFalse, f.Timeout)
+}
+
+// tests if the cluster is in orchestrator and is properly configured
+func testClusterIsInOrchestartor(f *framework.Framework, cluster *api.MysqlCluster, where string) {
+	By(fmt.Sprintf("Test cluster is in orchestrator: %s", where))
```

Comment:
These should be outside these helper functions, so the test reads more naturally.
```go
By("testing that cluster is registered with orchestrator")
testClusterIsRegistredWithOrchestrator()
```